File: 28数组中出现次数超过一半的数字.java
// Brute Force O(n^2)
// public class Solution {
//     public int MoreThanHalfNum_Solution(int [] array) {
//         if(array == null || array.length == 0) return 0;
        
//         int count;
//         for(int i = 0; i < array.length; i++) {
//             count = 0;
//             for(int j = 0; j < array.length; j++) {
//                 if(array[i] == array[j]) count++;
//                 if(count > array.length/2) {
//                     return array[j];
//                 }
//             }
//         }
//         return 0;
//     }
// }

// O(n)
public class Solution {
    public int MoreThanHalfNum_Solution(int[] array) {
        if(array == null || array.length == 0) return 0;
        int preVal = array[0];
        int count = 1;
        for(int i = 1; i < array.length; i++) {
            if(array[i] == preVal) count++;
            else {
                count--;
                if(count == 0) {
                    preVal = array[i];
                    count = 1;
                }
            }
        }
        int num = 0;
        for(int i = 0; i < array.length; i++) {
            if(array[i] == preVal) num ++;
        }
        return num > array.length/2 ? preVal : 0;
    }
}

File: 28数组中出现次数超过一半的数字.py
# Using dict
# -*- coding:utf-8 -*-
class Solution:
    def MoreThanHalfNum_Solution(self, numbers):
        # write code here
        if numbers is None:
            return None;
        length = len(numbers)
        if length == 0: return 0
        if length == 1: return numbers[0]
        
        numdict = {}
        for i in range(length):
            if(numbers[i] in numdict):
                numdict[numbers[i]] += 1
                if numdict[numbers[i]] > length//2:
                    return numbers[i]
            else:
                numdict[numbers[i]] = 1
        return 0

# Using sort More faster than above
# -*- coding:utf-8 -*-
class Solution:
    def MoreThanHalfNum_Solution(self, numbers):
        # write code here
        if numbers is None: return None
        length = len(numbers)
        if length == 0: return 0
        if length == 1: return numbers[0]
        
        numbers.sort()
        num = numbers[length // 2]
        if numbers.count(num) > length // 2:
            return num
        return 0